In [1]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

<h1>Data Investigation</h1>

I used the GapMinder data set to investigate the three variables incomeperperson, armedforcesrate, and polityscore.

<h4>SET UP</h4>

<i>Import the packages needed in this programme</i>

In [2]:
import pandas as pd
import numpy as np

<i>Set some options</i>

In [3]:
pd.set_option('display.max_rows', 220)
pd.set_option('expand_frame_repr', False)
pd.set_option('display.float_format',lambda x:'%f'%x)

<i>Read in the whole data set then set the index to be the Series country</i>

In [4]:
data = pd.read_csv('../gapminder.csv', low_memory=False, index_col='country')

In [5]:
for each in data.columns:
    data.loc[:,each] = pd.to_numeric(data.loc[:,each].str.replace(' ',''))

<h4>INVESTIGATION</h4>

<i>Look at the top 5 rows and the dimension of the DataFrame</i>

In [6]:
data.head(5)

incomeperperson  alcconsumption  armedforcesrate  breastcancerper100th      co2emissions  femaleemployrate  hivrate  internetuserate  lifeexpectancy  oilperperson  polityscore  relectricperperson  suicideper100th  employrate  urbanrate
country                                                                                                                                                                                                                                                 
Afghanistan              nan        0.030000         0.569653             26.800000   75944000.000000         25.600000      nan         3.654122       48.673000           nan     0.000000                 nan         6.684385   55.700001  24.040000
Albania          1914.996551        7.290000         1.024736             57.400000  223747333.333333         42.099998      nan        44.989947       76.918000           nan     9.000000          636.341383         7.699330   51.400002  46.720000
Algeria          2231.993335        0.690000         2.306817             23.500000 2932108666.666670         31.700001 0.100000        12.500073       73.131000      0.420095     2.000000          590.509814         4.848770   50.500000  65.220000
Andorra         21943.339898       10.170000              nan                   nan               nan               nan      nan        81.000000             nan           nan          nan                 nan         5.362179         nan  88.920000
Angola           1381.004268        5.570000         1.461329             23.100000  248358000.000000         69.400002 2.000000         9.999954       51.093000           nan    -2.000000          172.999227        14.554677   75.699997  56.700000

In [7]:
data.shape

(213, 15)

<i>Income per person variable</i>

In [8]:
c1 = data["incomeperperson"].value_counts(sort = False).sort_index()
p1 = data["incomeperperson"].value_counts(sort = False, normalize = True)
u1 = data["incomeperperson"].unique().size
n1 = data["incomeperperson"].isnull()[data["incomeperperson"].isnull() == True].size
ct1 = data.groupby(data['incomeperperson'],sort=True).size()*100/len(data)
fc1 = np.cumsum(c1)
fp1 = np.cumsum(ct1)
ippFreqTable = pd.concat([c1,p1,ct1,fc1,fp1],axis=1)
ippFreqTable.columns = ['Value Counts','Normalised Percentage','Percentage','Cumulative Value Counts','Cumulative Percentage']
ippFreqTable.index.name = 'incomeperperson'
valc1 = data['incomeperperson'].value_counts().sum()

In [9]:
print('The number of unique values is: ', u1)
print('The number of null values is: ',n1)
ippFreqTable.sort_values('Cumulative Value Counts').head()
ippFreqTable.sort_values('Cumulative Value Counts').tail()

The number of unique values is:  191
The number of null values is:  23


Value Counts  Normalised Percentage  Percentage  Cumulative Value Counts  Cumulative Percentage
incomeperperson                                                                                                 
103.775857                  1               0.005263    0.469484                        1               0.469484
115.305996                  1               0.005263    0.469484                        2               0.938967
131.796207                  1               0.005263    0.469484                        3               1.408451
155.033231                  1               0.005263    0.469484                        4               1.877934
161.317137                  1               0.005263    0.469484                        5               2.347418

Value Counts  Normalised Percentage  Percentage  Cumulative Value Counts  Cumulative Percentage
incomeperperson                                                                                                 
39972.352768                1               0.005263    0.469484                      186              87.323944
52301.587179                1               0.005263    0.469484                      187              87.793427
62682.147006                1               0.005263    0.469484                      188              88.262911
81647.100031                1               0.005263    0.469484                      189              88.732394
105147.437697               1               0.005263    0.469484                      190              89.201878

In [10]:
print('The number of coountries with GDP:')
print('less than or equal to $1,000 is:', data['incomeperperson'][data['incomeperperson'] < 1000].value_counts().sum(), 'which is',round(data['incomeperperson'][data['incomeperperson'] < 1000].value_counts().sum()*100/valc1,2),'%')
print('between $1,000 and $5,000 is:', data['incomeperperson'][(data['incomeperperson'] < 5000) &  (data['incomeperperson'] >= 1000)].value_counts().sum(),'which is', round(data['incomeperperson'][(data['incomeperperson'] < 5000) & (data['incomeperperson'] >= 1000)].value_counts().sum()*100/valc1,2),'%')
print('between $5,000 and $10,000 is:', data['incomeperperson'][(data['incomeperperson'] < 10000) & (data['incomeperperson'] >= 5000)].value_counts().sum(),'which is', round(data['incomeperperson'][(data['incomeperperson'] < 10000) & (data['incomeperperson'] >= 5000)].value_counts().sum()*100/valc1,2),'%')
print('between $10,000 and $25,000 is:', data['incomeperperson'][(data['incomeperperson'] < 25000) & (data['incomeperperson'] >= 10000)].value_counts().sum(),'which is', round(data['incomeperperson'][(data['incomeperperson'] < 25000) & (data['incomeperperson'] >= 10000)].value_counts().sum()*100/valc1,2),'%')    
print('between $25,000 and $50,000 is:', data['incomeperperson'][(data['incomeperperson'] < 50000) & (data['incomeperperson'] >= 25000)].value_counts().sum(),'which is', round(data['incomeperperson'][(data['incomeperperson'] < 50000) & (data['incomeperperson'] >= 25000)].value_counts().sum()*100/valc1,2),'%')    
print('greater than $50,000 is:', data['incomeperperson'][data['incomeperperson'] > 50000].value_counts().sum(), 'which is',round(data['incomeperperson'][data['incomeperperson'] > 50000].value_counts().sum()*100/valc1,2),'%')

The number of coountries with GDP:
less than or equal to $1,000 is: 54 which is 28.42 %
between $1,000 and $5,000 is: 61 which is 32.11 %
between $5,000 and $10,000 is: 28 which is 14.74 %
between $10,000 and $25,000 is: 23 which is 12.11 %
between $25,000 and $50,000 is: 20 which is 10.53 %
greater than $50,000 is: 4 which is 2.11 %


<i>Armed forces rate variable</i>

In [11]:
c2 = data["armedforcesrate"].value_counts(sort = False).sort_index()
p2 = data["armedforcesrate"].value_counts(sort = False, normalize = True)
u2 = data["armedforcesrate"].unique().size
n2 = data["armedforcesrate"].isnull()[data["armedforcesrate"].isnull() == True].size
ct2 = data.groupby(data['armedforcesrate'],sort=True).size()*100/len(data)
fc2 = np.cumsum(c2)
fp2 = np.cumsum(ct2)
ippFreqTable = pd.concat([c2,p2,ct2,fc2,fp2],axis=1)
ippFreqTable.columns = ['Value Counts','Normalised Percentage','Percentage','Cumulative Value Counts','Cumulative Percentage']
ippFreqTable.index.name = 'armedforcesrate'
valc2 = data['armedforcesrate'].value_counts().sum()

In [12]:
print('The number of unique values is: ', u2)
print('The number of null values is: ',n2)
ippFreqTable.sort_values('Cumulative Value Counts').head()
ippFreqTable.sort_values('Cumulative Value Counts').tail()

The number of unique values is:  165
The number of null values is:  49


Value Counts  Normalised Percentage  Percentage  Cumulative Value Counts  Cumulative Percentage
armedforcesrate                                                                                                 
0.000000                    1               0.006098    0.469484                        1               0.469484
0.066100                    1               0.006098    0.469484                        2               0.938967
0.102269                    1               0.006098    0.469484                        3               1.408451
0.105115                    1               0.006098    0.469484                        4               1.877934
0.114592                    1               0.006098    0.469484                        5               2.347418

Value Counts  Normalised Percentage  Percentage  Cumulative Value Counts  Cumulative Percentage
armedforcesrate                                                                                                 
5.955847                    1               0.006098    0.469484                      160              75.117371
6.394936                    1               0.006098    0.469484                      161              75.586854
7.737913                    1               0.006098    0.469484                      162              76.056338
9.820127                    1               0.006098    0.469484                      163              76.525822
10.638521                   1               0.006098    0.469484                      164              76.995305

In [13]:
print('The number of countries with armed forces rate:')
print('less than 0.5% of total work force:', data['armedforcesrate'][data['armedforcesrate'] < 0.5].value_counts().sum(), 'which is', round(data['armedforcesrate'][data['armedforcesrate'] < 0.5].value_counts().sum()*100/valc2,2),'%')
print('between 0.5% and 1%:', data['armedforcesrate'][(data['armedforcesrate'] < 1) & (data['armedforcesrate'] >= 0.5)].value_counts().sum(), 'which is',round(data['armedforcesrate'][(data['armedforcesrate'] < 1) & (data['armedforcesrate'] >= 0.5)].value_counts().sum()*100/valc2,2),'%')
print('between 1% and 5%:', data['armedforcesrate'][(data['armedforcesrate'] < 5) & (data['armedforcesrate'] >= 1)].value_counts().sum(),'which is', round(data['armedforcesrate'][(data['armedforcesrate'] < 5) & (data['armedforcesrate'] >= 1)].value_counts().sum()*100/valc2,2),'%')
print('between 5% and 7.5%:', data['armedforcesrate'][(data['armedforcesrate'] < 7.5) & (data['armedforcesrate'] >= 5)].value_counts().sum(),'which is', round(data['armedforcesrate'][(data['armedforcesrate'] < 7.5) & (data['armedforcesrate'] >= 5)].value_counts().sum()*100/valc2,2),'%')
print('greater than 7.5%:', data['armedforcesrate'][data['armedforcesrate'] > 7.5].value_counts().sum(), 'which is',round(data['armedforcesrate'][data['armedforcesrate'] > 7.5].value_counts().sum()*100/valc2,2),'%')

The number of countries with armed forces rate:
less than 0.5% of total work force: 43 which is 26.22 %
between 0.5% and 1%: 46 which is 28.05 %
between 1% and 5%: 65 which is 39.63 %
between 5% and 7.5%: 7 which is 4.27 %
greater than 7.5%: 3 which is 1.83 %


<i>Polity score variable</i>

In [14]:
c3 = data["polityscore"].value_counts(sort = False).sort_index()
p3 = data["polityscore"].value_counts(sort = False, normalize = True)
u3 = data["polityscore"].unique().size
n3 = data["polityscore"].isnull()[data["polityscore"].isnull() == True].size
ct3 = data.groupby(data['polityscore'],sort=True).size()*100/len(data)
fc3 = np.cumsum(c3)
fp3 = np.cumsum(ct3)
ippFreqTable = pd.concat([c3,p3,ct3,fc3,fp3],axis=1)
ippFreqTable.columns = ['Value Counts','Normalised Percentage','Percentage','Cumulative Value Counts','Cumulative Percentage']
ippFreqTable.index.name = 'polityscore'

In [15]:
print('The number of unique values is: ', u3)
print('The number of null values is: ',n3)
ippFreqTable.sort_values('Cumulative Value Counts')

The number of unique values is:  22
The number of null values is:  52


Value Counts  Normalised Percentage  Percentage  Cumulative Value Counts  Cumulative Percentage
polityscore                                                                                                 
-10.000000              2               0.012422    0.938967                        2               0.938967
-9.000000               4               0.024845    1.877934                        6               2.816901
-8.000000               2               0.012422    0.938967                        8               3.755869
-7.000000              12               0.074534    5.633803                       20               9.389671
-6.000000               3               0.018634    1.408451                       23              10.798122
-5.000000               2               0.012422    0.938967                       25              11.737089
-4.000000               6               0.037267    2.816901                       31              14.553991
-3.000000               6               0.037267    2.816901                       37              17.370892
-2.000000               5               0.031056    2.347418                       42              19.718310
-1.000000               4               0.024845    1.877934                       46              21.596244
0.000000                6               0.037267    2.816901                       52              24.413146
1.000000                3               0.018634    1.408451                       55              25.821596
2.000000                3               0.018634    1.408451                       58              27.230047
3.000000                2               0.012422    0.938967                       60              28.169014
4.000000                4               0.024845    1.877934                       64              30.046948
5.000000                7               0.043478    3.286385                       71              33.333333
6.000000               10               0.062112    4.694836                       81              38.028169
7.000000               13               0.080745    6.103286                       94              44.131455
8.000000               19               0.118012    8.920188                      113              53.051643
9.000000               15               0.093168    7.042254                      128              60.093897
10.000000              33               0.204969   15.492958                      161              75.586854

<h4>DISCUSSION</h4>

I chose not to print out all the rows of two of my variable frequency tables (indicated in the printed lines) because the number of rows was so large, but the whole table still exists in the memory so it can be looked at fully if needed. All my percentages and cumulative percentages include the null values.

For incomeperperson there were so many different unique values that I decided to quickly proportion it off and look at some groupings of it that printed out below the main table. From this I can see that just over three quarters of the total is in the $0 to $10,000 GDP range, for me this is slightly surprising as I thought the proportion would be large in that range but not quite that large. The values are 2010 GDP converted to constant year 2000 US$ and are not modified for different costs of living in different countries and I think the huge number of people in that lower bracket is mostly a reflection on the lower costs of living in those countries.

For armedforcesrate again there were so many unique values that I decided to show only the first and last few in the output but the whole table is in memory and in looking at the whole I could choose my brackets like with the incomeperperson variable. Having then bracketed and summarised the data a little I could see that nearly 55% of the countries have 1% or less percentage of the active work force in the armed forces. This didn’t surprise me, nor does just over 6% of the countries having a 5% or higher proportion of the work force in the armed forces. 

Finally, for the polityscore it was easy to fit all the values in to the view. I was surprised that quite so many countries were up in the top levels of freedom, over 15% are ranked in the top grade alone and, given so many countries seem to be overthrowing their governments recently, under 15% of countries fall into the -10 to -4 score.  It’s interesting that so many countries don’t have a score, and that is due to the researchers choice, but it could also be the reason we have so many countries in the higher grades and fewer in the lower.